audiobook/src/main/java/de/ph1b/audiobook/features/chapterReader/oggReading.kt

```diff
@@ -0,0 +1,165 @@
+package de.ph1b.audiobook.features.chapterReader
+
+import android.util.SparseArray
+import java.io.EOFException
+import java.io.InputStream
+import java.util.ArrayDeque
+
+private val OGG_PAGE_MAGIC = "OggS".toByteArray()
+
+internal data class OggPage(
+    val continuedPacket: Boolean,
+    val finishedPacket: Boolean,
+    val firstPageOfStream: Boolean,
+    val lastPageOfStream: Boolean,
+    val absoluteGranulePosition: Long,
+    val streamSerialNumber: Int,
+    val pageSequenceNumber: Long,
+    val packets: List<ByteArray>) {
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is OggPage) return false
+    return continuedPacket == other.continuedPacket &&
+        finishedPacket == other.finishedPacket &&
+        firstPageOfStream == other.firstPageOfStream &&
+        lastPageOfStream == other.lastPageOfStream &&
+        absoluteGranulePosition == other.absoluteGranulePosition &&
+        streamSerialNumber == other.streamSerialNumber &&
+        pageSequenceNumber == other.pageSequenceNumber &&
+        packets.size == other.packets.size &&
+        packets.indices.all { packets[it] contentEquals other.packets[it] }
+  }
+
+  override fun hashCode()
+      = listOf(continuedPacket.hashCode(), finishedPacket.hashCode(), firstPageOfStream.hashCode(),
```

Comment:
One argument (nevertheless, quite weak in this case) might be that it's just a better hash function because addition is commutative so hash function based on it doesn't take the order of fields into account. For example:
```
>>> val l = (0..255).map { v -> (0..7).map { (v and (1 shl it) != 0).hashCode() } }
>>> l.map { it.reduce { acc, i -> acc + i } }.toSet().size
9
>>> l.map { it.hashCode() }.toSet().size
256
```